File: Algorithms/173. Binary Search Tree Iterator/swift/solution.flattening.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */

class BSTIterator {

    var nodeSorted = [Int]()
    var index = -1

    init(_ root: TreeNode?) {
        inorder(root)
    }

    func inorder(_ node: TreeNode?) {
        guard let node = node else {
            return
        }

        inorder(node.left)
        nodeSorted.append(node.val)
        inorder(node.right)
    }

    func next() -> Int {
        index += 1
        return nodeSorted[index]
    }

    func hasNext() -> Bool {
        return index + 1 < nodeSorted.count
    }
}


/**
 * Your BSTIterator object will be instantiated and called as such:
 * let obj = BSTIterator(root)
 * let ret_1: Int = obj.next()
 * let ret_2: Bool = obj.hasNext()
 */

File: Algorithms/173. Binary Search Tree Iterator/php/solution.flattening.php
<?php

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class BSTIterator
{
    private $nodesSorted = array();
    private $index = -1;

    /**
     * @param TreeNode $root
     */
    function __construct($root)
    {
        $this->inorder($root);
    }

    function inorder($root)
    {
        if (is_null($root)) {
            return;
        }
        $this->inorder($root->left);
        $this->nodesSorted[] = $root->val;
        $this->inorder($root->right);
    }

    /**
     * @return Integer
     */
    function next()
    {
        return $this->nodesSorted[++$this->index];
    }

    /**
     * @return Boolean
     */
    function hasNext()
    {
        return $this->index + 1 < count($this->nodesSorted);
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * $obj = BSTIterator($root);
 * $ret_1 = $obj->next();
 * $ret_2 = $obj->hasNext();
 */
